File: src/main/java/org/goodmath/pica/ast/Identifier.java
/* Copyright 2022, Mark C. Chu-Carroll
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.goodmath.pica.ast;

import lombok.EqualsAndHashCode;
import org.goodmath.pica.ast.locations.Location;
import org.goodmath.pica.util.Twist;

import java.util.List;
import java.util.Optional;

@EqualsAndHashCode(callSuper = false)
public class Identifier extends AstNode {
    private final Optional<Identifier> moduleId;
    private final String name;

    public Identifier(Optional<Identifier> moduleId, String name, Location loc) {
        super(loc);
        this.moduleId = moduleId;
        this.name = name;
    }

    public Optional<Identifier> getModule() {
        return moduleId;
    }

    public String getName() {
        return name;
    }

    @Override
    public String toString() {
        if (getModule().isPresent()) {
            return getModule().get() + "::" + getName();
        } else {
            return getName();
        }
    }

    public static Identifier parseIdentifier(String idStr) {
        String[] bits = idStr.split("::");
        return fromList(List.of(bits));
    }

    public static Identifier fromList(List<String> bits) {
        Identifier current = null;
        for (String bit: bits) {
            current = new Identifier(Optional.ofNullable(current),
                bit, Location.NO_LOCATION);
        }
        return current;
    }

    @Override
    public Twist twist() {
        return Twist.obj("Identifier",
                Twist.attr("name", toString()));

    }


}

File: src/main/kotlin/org/goodmath/pica/ast/Identifier.kt
package org.goodmath.pica.ast

import org.goodmath.pica.util.Symbol
import org.goodmath.pica.util.twist.Twist

class Identifier(val hadronId: Identifier?, val name: Symbol, loc: Location): AstNode(loc) {
    override fun twist(): Twist =
        if (hadronId == null) {
            Twist.obj(
                "Ident",
                Twist.attr("name", name.repr)
            )
        } else {
            Twist.obj(
                "Ident",
                Twist.value("hadron", hadronId),
                Twist.attr("name", name.repr)
            )

    }

    override fun toString(): String {
        return if (hadronId != null) {
            "$hadronId::${name.repr}"
        } else {
            name.repr
        }
    }

    companion object {
        fun fromList(parts: List<String>, loc: Location): Identifier {
            var id: Identifier? = null
            for (p in parts) {
                id = Identifier(id, Symbol.fromString(p), loc)
            }
            return id!!
        }

        fun fromString(id: String, loc: Location): Identifier {
            return fromList(id.split("::"), loc)
        }
    }
}